Use the workspace trust feature

shati-patel opened the pull request:
Will fix https://github.com/github/vscode-codeql/issues/849 and will also (indirectly) fix https://github.com/github/vscode-codeql/issues/509. 

Uses the syntax from https://github.com/microsoft/vscode/issues/120251#issuecomment-825832603 to let the CodeQL extension work with workspace trust. For users who have enabled the workspace trust feature:
- If the user trusts the workspace: all CodeQL features are enabled
- If the user doesn't trust the workspace: Certain settings (that could execute arbitrary code from the workspace) are disabled at the workspace level, and the extension ignores them. This is what the UI looks like:
![image](https://user-images.githubusercontent.com/42641846/118016834-c68e4b80-b34d-11eb-8f5b-41607bab4fa9.png)

⚠️ ~Note: Currently, you can't use the CodeQL extension at all in untrusted workspaces. This is because it depends on the [Test Explorer](https://marketplace.visualstudio.com/items?itemName=hbenl.vscode-test-explorer), which currently doesn't work in untrusted workspaces. PR: https://github.com/hbenl/vscode-test-explorer/pull/204.~ Fixed!

## Checklist

- [x] Once VS Code is released with workspace trust enabled by default, we need to update the minimum VS Code version required to run this extension. See: 
https://github.com/github/vscode-codeql/blob/72d57eec6ecfb6ff10d59a64d8403acb68ef05b5/extensions/ql-vscode/package.json#L15-L17
- [x] [CHANGELOG.md](https://github.com/github/vscode-codeql/blob/main/extensions/ql-vscode/CHANGELOG.md) has been updated to incorporate all user visible changes made by this pull request.
- [x] Issues have been created for any UI or other user-facing changes made by this pull request.
- [x] `@github/docs-content-codeql` has been cc'd in all issues for UI or other user-facing changes made by this pull request.
  - Will require small update to https://codeql.github.com/docs/codeql-for-visual-studio-code/customizing-settings:
    > To override the default behavior and use a different CLI, you can specify the CodeQL CLI Executable Path. ~Note that this is only available as a user setting, not as a workspace setting.~ 
  - Docs PR opened here: https://github.com/github/codeql/pull/6072

aeisenberg commented:
This looks great! Two extra things:

1. This PR should also address #509 because users can now specify per-workspace cli versions.
2. We should not merge this until workspace trust is enabled by default in vscode. I'm not sure which version that will be, but I think it's happening soon.
3. We need to update the minimum vscode version required to run this extension when the appropriate vscode is released. See: 
https://github.com/github/vscode-codeql/blob/72d57eec6ecfb6ff10d59a64d8403acb68ef05b5/extensions/ql-vscode/package.json#L15-L17

shati-patel commented:
> The upstream PR has been merged and the new version of test explorer has been released. Let's ship this!

We should probably still wait until until workspace trust is enabled by default in VS Code, since we've changed the CLI executable setting [back to "window" scope](https://github.com/github/vscode-codeql/pull/861/files#diff-be4c5a31505df7de81fa4900e84e66c04d1a6bca9b3da50f8061d2cb2bfbbde1R131). (i.e. anyone who hasn't explicitly set `"security.workspace.trust.enabled": true` doesn't yet get any of the protection offered by workspace trust.)

Points 2 and 3 in your previous comment (https://github.com/github/vscode-codeql/pull/861#issuecomment-839962213) still hold, I think!

adityasharad commented:
Nice! Do we know what version of VS Code will ship with workspace trust?

shati-patel commented:
> Nice! Do we know what version of VS Code will ship with workspace trust?

I've been keeping an eye on https://github.com/microsoft/vscode/issues/120251, but haven't seen anything about an expected release. (Maybe I'm looking in the wrong places though...)

I suppose we could revert the "machine -> window" change and ship the workspace trust implementation itself at any time? That would at least make the extension work for users who _have_ enabled workspace trust and are in an untrusted workspace, without changing the behaviour for anyone else.

aeisenberg commented:
> We should probably still wait until until workspace trust is enabled by default in VS Code

Yes. Good point. 

There is a block on the package.json and we should update this to the version where workspace trust is going to be released. This will ensure that users on older versions on vscode won't accidentally get the new configuration behaviour.

```json
"engines": {
    "vscode": "^1.43.0"
},
```

alexet commented:
Running qtests in untrusted workspaces is also unsafe at the moment and hence also needs to be disabled.

shati-patel commented:
I've added some more commits to update dependencies, since workspace trust is now fully enabled (v1.57.0). I updated to 1.48.0  separately first, using commits from https://github.com/github/vscode-codeql/pull/874 (which will hopefully minimise conflicts with the `qc-development` branch), so that commit (https://github.com/github/vscode-codeql/pull/861/commits/ec4785ea7e6ef11b6baff7c60135097614591784) shouldn't need a re-review.




------
Re:
> Running qtests in untrusted workspaces is also unsafe at the moment and hence also needs to be disabled.

I think that's what line 30 covers: https://github.com/github/vscode-codeql/blob/85a1f6663bd1c5ac382ad64a2c4f73301f0d55df/extensions/ql-vscode/package.json#L28-L31

(Unless we need to disable qltests completely? I'm not sure how to do that.)

aeisenberg commented on extensions/ql-vscode/CHANGELOG.md:
```diff
@@ -6,6 +6,7 @@
 - Display a helpful message and link to documentation when a query produces no results. [#866](https://github.com/github/vscode-codeql/pull/866)
 - Refresh test databases automatically after a test run. [#868](https://github.com/github/vscode-codeql/pull/868)
 - Allow users to specify a custom directory for storing query server logs (`codeQL.runningQueries.customLogDirectory`). The extension will not delete these logs automatically. [#863](https://github.com/github/vscode-codeql/pull/863)
+- Support the "workspace trust" feature. This extension is enabled in untrusted workspaces, but it restricts commands that contain arbitrary paths. [#861](https://github.com/github/vscode-codeql/pull/861)
```
```suggestion
- Support the VS Code [Workspace Trust feature](https://code.visualstudio.com/docs/editor/workspace-trust). This extension is now enabled in untrusted workspaces, but it restricts commands that contain arbitrary paths. [#861](https://github.com/github/vscode-codeql/pull/861)
- Allow the `codeQL.cli.executablePath` configuration setting to be set in worskpace-scoped configuration files.  This means that each workspace can now specify its own CLI compiler, a feature that is unblocked due to implementing Workspace Trust. [#861](https://github.com/github/vscode-codeql/pull/861)
```

shati-patel replied:
Thanks for adding the extra context!

aeisenberg commented on extensions/ql-vscode/package.json:
```diff
@@ -13,14 +13,24 @@
+  "capabilities": {
+    "untrustedWorkspaces": {
+      "supported": "limited",
+      "description": "Workspace trust is required to execute commands that can contain arbitrary paths. ",
```
```suggestion
      "description": "Workspace trust is required to execute commands that can contain arbitrary paths.",
```

aeisenberg reviewed (commented):
Looks good. Just some updates for the changelog.
